serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,472 @@
+        list_logit_bias_indices: List[List[int]],
+        list_logit_bias_values: List[List[float]],
+        list_past_output_tokens: List[List[int]],
+    ):
+        # NOTE: Keep `mask_random_t` and `mask_greedy_t` tensors in CPU.
+        #       Moving them to gpu showed a small performance regression.
+        mask_random = torch.tensor(
+            list_mask_random,
+            dtype=torch.bool,
+            device="cpu",
+        )
+        mask_greedy = torch.logical_not(
+            mask_random,
+        )
+        # `mask_top_logprob` will be on cpu
+        mask_top_logprob = torch.from_numpy(list_mask_top_logprob)
+        temp = torch.tensor(
+            list_temperatures,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        top_ps = torch.tensor(
+            list_top_ps,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        top_ks = torch.tensor(
+            list_top_ks,
+            dtype=torch.int,
+            device="cpu",
+            pin_memory=True,
+        )
+        frequency_penalties = torch.tensor(
+            list_frequency_penalties,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        presence_penalties = torch.tensor(
+            list_presence_penalties,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        repetition_penalties = torch.tensor(
+            list_repetition_penalties,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        logit_bias_indices = torch.tensor(
+            list_logit_bias_indices,
+            dtype=torch.long,
+            device="cpu",
+            pin_memory=True,
+        )
+        # Convert 1-based index to 0-based
+        logit_bias_indices -= 1
+        logit_bias_values = torch.tensor(
+            list_logit_bias_values,
+            dtype=dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+        past_output_tokens = torch.tensor(
+            list_past_output_tokens,
+            dtype=torch.long,
+            device="cpu",
+            pin_memory=True,
+        )
+
+        return cls(
+            mask_random,
+            mask_greedy,
+            mask_top_logprob,
+            temp.to(device=dev, non_blocking=True),
+            top_ps.to(device=dev, non_blocking=True),
+            top_ks.to(device=dev, non_blocking=True),
+            frequency_penalties.to(device=dev, non_blocking=True),
+            presence_penalties.to(device=dev, non_blocking=True),
+            repetition_penalties.to(device=dev, non_blocking=True),
+            logit_bias_indices.to(device=dev, non_blocking=True),
+            logit_bias_values.to(device=dev, non_blocking=True),
+            past_output_tokens.to(device=dev, non_blocking=True),
+        )
+
+
+@dataclass
+class SamplingMetadata:
```

Comment:
Isn't this technically the sampler state? naming it "metadata" makes it feel like its something else, can we also put a doc comment on the class and all the fields? if you want to do as follow up let's file tickets.